File: fundExpert/tests/example_client.py
"""
示例：如何在 Python 中调用基金分析服务
"""
import requests
from typing import List, Dict


class FundAnalysisClient:
    """基金分析服务客户端"""
    
    def __init__(self, base_url: str = "http://localhost:8001"):
        """
        初始化客户端
        
        Args:
            base_url: 服务地址
        """
        self.base_url = base_url
        self.api_base = f"{base_url}/api/v1"
    
    def health_check(self) -> Dict:
        """
        健康检查
        
        Returns:
            健康状态
        """
        response = requests.get(f"{self.api_base}/health")
        response.raise_for_status()
        return response.json()
    
    def analyze_portfolio(self, holdings: List[Dict]) -> Dict:
        """
        分析投资组合
        
        Args:
            holdings: 持仓列表
            
        Returns:
            分析结果
            
        Example:
            holdings = [
                {
                    "fund_code": "110022",
                    "amount": 30000,
                    "ratio": 0.30,
                    "asset_type": "stock"
                }
            ]
        """
        payload = {"holdings": holdings}
        response = requests.post(
            f"{self.api_base}/analyze",
            json=payload,
            headers={"Content-Type": "application/json"}
        )
        response.raise_for_status()
        return response.json()


# 使用示例
if __name__ == "__main__":
    # 创建客户端
    client = FundAnalysisClient()
    
    # 检查服务状态
    print("检查服务状态...")
    health = client.health_check()
    print(f"服务状态: {health['status']}")
    print()
    
    # 准备持仓数据
    holdings = [
        {
            "fund_code": "110022",  # 易方达消费行业
            "amount": 30000,
            "ratio": 0.30,
            "asset_type": "stock"
        },
        {
            "fund_code": "000088",  # 嘉实中证金边
            "amount": 40000,
            "ratio": 0.40,
            "asset_type": "long_bond"
        },
        {
            "fund_code": "000057",  # 中银消费主题
            "amount": 15000,
            "ratio": 0.15,
            "asset_type": "mid_bond"
        },
        {
            "fund_code": "160216",  # 国泰商品
            "amount": 7500,
            "ratio": 0.075,
            "asset_type": "commodity"
        },
        {
            "fund_code": "518880",  # 黄金ETF
            "amount": 7500,
            "ratio": 0.075,
            "asset_type": "gold"
        }
    ]
    
    # 分析投资组合
    print("分析投资组合...")
    print("持仓数据:", holdings)
    print()
    
    try:
        result = client.analyze_portfolio(holdings)
        
        print("分析结果:")
        print(f"状态: {result['status']}")
        print(f"时间: {result['timestamp']}")
        print()
        
        analysis = result['analysis']
        print("分析摘要:", analysis['summary'])
        print()
        
        print("建议:")
        for rec in analysis['recommendations']:
            print(f"  - {rec}")
        print()
        
        if analysis['risk_warnings']:
            print("风险提示:")
            for warning in analysis['risk_warnings']:
                print(f"  ⚠️  {warning}")
        
        print()
        print("完整报告:")
        print(analysis['raw_output'])
        
    except requests.exceptions.RequestException as e:
        print(f"请求失败: {str(e)}")
    except Exception as e:
        print(f"错误: {str(e)}")

File: fundExpert/example_client.ts
/**
 * 基金分析服务客户端（TypeScript/JavaScript）
 * 用于前端调用后端分析服务
 */

interface Holding {
  fund_code: string;
  amount: number;
  ratio: number;
  asset_type: 'stock' | 'long_bond' | 'mid_bond' | 'commodity' | 'gold';
}

interface AnalysisRequest {
  holdings: Holding[];
}

interface AnalysisResponse {
  status: string;
  timestamp: string;
  holdings: Holding[];
  analysis: {
    raw_output: string;
    summary: string;
    allocation_analysis: {
      description: string;
      has_deviation: boolean;
    };
    recommendations: string[];
    risk_warnings: string[];
  };
}

interface HealthResponse {
  status: string;
  service: string;
  version: string;
}

class FundAnalysisClient {
  private baseUrl: string;
  private apiBase: string;

  constructor(baseUrl: string = 'http://localhost:8001') {
    this.baseUrl = baseUrl;
    this.apiBase = `${baseUrl}/api/v1`;
  }

  /**
   * 健康检查
   */
  async healthCheck(): Promise<HealthResponse> {
    const response = await fetch(`${this.apiBase}/health`);
    if (!response.ok) {
      throw new Error(`健康检查失败: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * 分析投资组合
   */
  async analyzePortfolio(holdings: Holding[]): Promise<AnalysisResponse> {
    const response = await fetch(`${this.apiBase}/analyze`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ holdings }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || `分析失败: ${response.statusText}`);
    }

    return response.json();
  }
}

// ===== 使用示例 =====

// 示例1：基础使用
async function example1() {
  const client = new FundAnalysisClient();

  // 健康检查
  const health = await client.healthCheck();
  console.log('服务状态:', health.status);

  // 准备持仓数据
  const holdings: Holding[] = [
    {
      fund_code: '110022',
      amount: 30000,
      ratio: 0.3,
      asset_type: 'stock',
    },
    {
      fund_code: '000088',
      amount: 40000,
      ratio: 0.4,
      asset_type: 'long_bond',
    },
    {
      fund_code: '000057',
      amount: 15000,
      ratio: 0.15,
      asset_type: 'mid_bond',
    },
    {
      fund_code: '160216',
      amount: 7500,
      ratio: 0.075,
      asset_type: 'commodity',
    },
    {
      fund_code: '518880',
      amount: 7500,
      ratio: 0.075,
      asset_type: 'gold',
    },
  ];

  // 分析
  const result = await client.analyzePortfolio(holdings);
  console.log('分析结果:', result);
  console.log('建议:', result.analysis.recommendations);
}

// 示例2：在 Vue 3 中使用
// import { ref } from 'vue';
//
// export function useFundAnalysis() {
//   const client = new FundAnalysisClient();
//   const loading = ref(false);
//   const result = ref<AnalysisResponse | null>(null);
//   const error = ref<string | null>(null);
//
//   const analyze = async (holdings: Holding[]) => {
//     loading.value = true;
//     error.value = null;
//     try {
//       result.value = await client.analyzePortfolio(holdings);
//     } catch (e: any) {
//       error.value = e.message;
//     } finally {
//       loading.value = false;
//     }
//   };
//
//   return {
//     loading,
//     result,
//     error,
//     analyze,
//   };
// }

// 示例3：在 React 中使用
// import { useState } from 'react';
//
// export function useFundAnalysis() {
//   const [loading, setLoading] = useState(false);
//   const [result, setResult] = useState<AnalysisResponse | null>(null);
//   const [error, setError] = useState<string | null>(null);
//
//   const client = new FundAnalysisClient();
//
//   const analyze = async (holdings: Holding[]) => {
//     setLoading(true);
//     setError(null);
//     try {
//       const data = await client.analyzePortfolio(holdings);
//       setResult(data);
//     } catch (e: any) {
//       setError(e.message);
//     } finally {
//       setLoading(false);
//     }
//   };
//
//   return {
//     loading,
//     result,
//     error,
//     analyze,
//   };
// }

// 示例4：错误处理
async function example4() {
  const client = new FundAnalysisClient();

  try {
    const holdings: Holding[] = [
      {
        fund_code: '110022',
        amount: 30000,
        ratio: 0.3,
        asset_type: 'stock',
      },
    ];

    const result = await client.analyzePortfolio(holdings);
    console.log('分析成功:', result);
  } catch (error: any) {
    console.error('分析失败:', error.message);
    // 可以根据错误类型做不同处理
    if (error.message.includes('网络')) {
      console.log('请检查网络连接');
    } else if (error.message.includes('服务')) {
      console.log('服务暂时不可用，请稍后重试');
    }
  }
}

// 示例5：在 Axios 环境中使用
// import axios from 'axios';
//
// class FundAnalysisClientAxios {
//   private apiBase: string;
//
//   constructor(baseUrl: string = 'http://localhost:8001') {
//     this.apiBase = `${baseUrl}/api/v1`;
//   }
//
//   async analyzePortfolio(holdings: Holding[]): Promise<AnalysisResponse> {
//     const response = await axios.post(`${this.apiBase}/analyze`, {
//       holdings,
//     });
//     return response.data;
//   }
// }

// 导出
export { FundAnalysisClient, type Holding, type AnalysisResponse, type HealthResponse };
